### PR TITLE
GTest support, Testing Framework, Continuous Integration and Code Coverage

### PR DESCRIPTION
**Pull request by jsrehak**
This pull request includes a GTest Framework for BART that allows us to easily use dealiilog feature to create and compare test files to provided gold files.

This also provides Travis CI support and Codecov.

Addresses #64  and #65

### CODE REVIEW

File: src/main.cc [hunk 1]
```diff
@@ -1,6 +1,51 @@
...
+  const struct option longopts[] =
+  {
+    {"report", no_argument, nullptr, 'r'}
+  };
+  
+  int c;
+  while ((c = getopt_long (argc, argv, "rd:", longopts, &option_index)) != -1) {
+    switch(c) {
+      case 'r':
+        btest::GlobalBartTestHelper().SetReport(true);
```
Comment (weixiong-zheng-berkeley) [inline src/main.cc]: I would suggest you change function and class naming with `Bart` to `BART` as each letter is actually the first letter of a word. Similar case is we have `ProduceAQ`

File: src/aqdata/aq_base.h [hunk 2]
```diff
@@ -5,7 +5,7 @@
 #include <map>
 #include <unordered_map>
 #include <string>
-
+#include <iomanip>
 #include <deal.II/base/tensor.h>
```
Comment (weixiong-zheng-berkeley) [inline src/aqdata/aq_base.h]: Add a line between 8 and 9 as `iomanip` is STL while deal.II is third-party library.

File: src/main.cc (hunk 1 above)
Comment (jsrehak) [reply src/main.cc]: I've updated that, as well as updating the class to `BARTTestHelper()`

File: src/aqdata/tests/lsgc_test.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
+#include "gtest/gtest.h"
```
Comment (weixiong-zheng-berkeley) [inline src/aqdata/tests/lsgc_test.cc]: `#include` should be subject to the following ordering: project headers->STL headers->third-party headers with one empty line separating each block. gtest is third-party so it should be put to the bottom

File: src/main.cc [hunk 4]
```diff
@@ -1,6 +1,51 @@
-#include "aqdata/aq_base.h"
+#ifdef TEST
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
```
Comment (weixiong-zheng-berkeley) [inline src/main.cc]: Same thing as above.

File: src/test_helpers/tests/bart_test_helper_test.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
+#include "gtest/gtest.h"
```
Comment (weixiong-zheng-berkeley) [inline src/test_helpers/tests/bart_test_helper_test.cc]: Same thing as above about ordering of include

File: src/test_helpers/bart_test_helper.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+(2) Running a gold file test to compare two files.
+ 
+ \author Joshua Rehak
+ \date 2018/2
+ */
+class BartTestHelper {
+ public:
+  //! Default constructor, default values `report = true`, `gold_files_directory = "test_data/"`
+  BartTestHelper();
+  //! Constructor specifying if a report will be generated and location of gold files.
```
Comment (weixiong-zheng-berkeley) [inline src/test_helpers/bart_test_helper.h]: An empty line separating each declaration would be better for readability. Human eyes are not good at looking at things staying too close.

File: src/test_helpers/bart_test_helper.h [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  void OpenLog(std::string filename);
+  
+  //! Detaches and closes the ofstream attached to the dealiilog
+  void CloseLog();
+  
+  //! Returns the status of the dealii log
+  bool IsLogging() const;
+  
+ private:
+  //! Used to clean up files following a gold test
```
Comment (weixiong-zheng-berkeley) [inline src/test_helpers/bart_test_helper.h]: Same thing here for empty lines between declarations.

File: src/test_helpers/tests/integration_test.cc [hunk 8]
```diff
@@ -0,0 +1,126 @@
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+#include <deal.II/base/utilities.h>
+#include <deal.II/base/logstream.h>
+
+#include <sys/stat.h>
+#include <exception>
+#include <fstream>
+
```
Comment (weixiong-zheng-berkeley) [inline src/test_helpers/tests/integration_test.cc]: `#include` ordering here.

**Pull request merged**